### PR TITLE
TAS: support rank-based ordering for JobSet

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Part of #3450 
Part of #3533 

#### Special notes for your reviewer:

I don't have e2e test yet, but getting early feedback would be great.

#### Does this PR introduce a user-facing change?

```release-note
TAS: support rank-based ordering for JobSet
```

### CODE REVIEW

**Comment by mimowo** [comment]
/assign @PBundyra 
cc @tenzen-y

File: pkg/controller/tas/topology_ungater.go [hunk 1]
```diff
@@ -396,39 +403,92 @@ func assignGatedPodsToDomainsGreedy(
...
-			return nil, false
+		rank := *podIndex
+		if rjInfo != nil {
+			jobIndex, err := readIntFromLabel(pod, rjInfo.jobIndexLabel)
+			if err != nil {
+				// the Pod has no Job index information - ranks cannot be used
+				return nil, err
+			}
+			singleJobSize := podSetSize / rjInfo.replicasCount
+			if *podIndex >= singleJobSize {
```
Comment (PBundyra) [inline pkg/controller/tas/topology_ungater.go:461]:
```suggestion
			if rank >= singleJobSize {
```

Comment (mimowo) [reply pkg/controller/tas/topology_ungater.go:461]: This would work, but isn't the podIndex better reflect the intention? The rank can still be mutated

**Comment by mimowo** [comment]
@PBundyra I have updated the e2e test and it is passing now, PTAL

File: test/e2e/tas/jobset_test.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+						},
+					},
+				).
+				Request("replicated-job-1", extraResource, "1").
+				Limit("replicated-job-1", extraResource, "1").
+				Obj()
+			sampleJob.Annotations = map[string]string{constants.QueueLabel: localQueue.Name}
+			gomega.Expect(k8sClient.Create(ctx, sampleJob)).Should(gomega.Succeed())
+
+			ginkgo.By("JobSet is unsuspended, and has all Pods active and ready", func() {
```
Comment (PBundyra) [inline test/e2e/tas/jobset_test.go:120]:
```suggestion
			ginkgo.By("JobSet is unsuspended", func() {
```

**Comment by PBundyra** [comment]
Thanks @mimowo!
/lgtm
/hold to address nits

**Comment by mimowo** [comment]
FYI. I have ran this test locally in a loop around 20 times and all passed.

**Comment by PBundyra** [comment]
/lgtm
Thank you

**Comment by mimowo** [comment]
/test pull-kueue-test-multikueue-e2e-main
Opened: https://github.com/kubernetes-sigs/kueue/issues/3600

**Pull request merged**